src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs

```diff
@@ -311,107 +440,85 @@ instance NoUnexpectedThunks a => NoUnexpectedThunks (Maybe a)
 instance NoUnexpectedThunks a => NoUnexpectedThunks (NonEmpty a)
 
 {-------------------------------------------------------------------------------
-  Using the standard 'isNormalForm' check
+  Spine-strict container types
+
+  Such types can /only/ contain thunks in the values, so that's all we check.
+  Note that containers using keys are typically strict in those keys, but that
+  forces them to WHNF only, not NF; in /most/ cases the @Ord@ instance on those
+  keys will force them to NF, but not /always/ (for example, when using lists
+  as keys).
```

Comment:
So we don't check for thunks in keys?

Reply:
No, no, we _do_. Let me say that explicitly.

Reply:
After discussing this on slack, introduced `noUnexpectedThunksInValues` and `noUnexpectedThunksInKeysAndValues`.